Appveyor CI enablement

tolbertam opened the pull request:
- Enables Appveyor CI for running integration tests against windows.  Initial goal is to run 'short' suite against jdk8 w/ C\* 3.0.5, 'unit' suites against jdk6 and jdk7.  Takes about ~50 minutes and should run rather reliably.
- Eventually to enable daily or weekly full suite runs against C\* 2.2+ once more confidence is established.
- Provides validation in ensuring that code that requires native os calls (netty-tcnative, jnr, etc.) works on windows.  Prebuilds libapr (if not present) and uses provided openssl for validating netty-tcnative.

Changes:
- Use powershell and use `--quiet-windows` flag on start (new C\* versions require this)
- Some platform specific quoting adjustments since windows is fickle about spaces.
- Some refactoring in `HostConnectionPoolTest` using `ConditionChecker` to ensure pools grow/shrink.
- Mark some tests as long that should be as they require multi-node clusters or take a long time.
- Add `exactRequired` field to `CassandraVersion` and `DseVersion` annotations to indicate that a test requires a specific version of Cassandra to run.
- Upgrade netty-tcnative test dependency from 1.1.33.Fork17 to 1.1.33.Fork18 as Fork17 does not work on windows.
- Upgrade scassandra to 1.0.9 (reduced thread count improves test stability on windows).

tolbertam committed:
Add appveyor continuous integration support

Enables continuous integration of builds against a windows platform.
'short' profile test runs will now be publically visible for pull
requests against the latest C* version and JDK8.  Unit tests run for
jdk6 and jdk7.

tolbertam committed:
Enhance CCMBridge to better support Windows

Use powershell instead of cmd.

Quotes jvmArgs as windows is a bit more picky about quotes.

Use --quiet-windows on start, this is required for C* versions post
2.2.4* to limit stderr/stdout output from C* on windows.

Also upgrades scassandra to 1.0.9 for better windows support.  See:
https://github.com/scassandra/scassandra-server/pull/157 for more
details.

tolbertam committed:
Mark SessionLeakTest as long since it bootstraps

SessionLeakTest bootstraps a node into an existing cluster which can
take a while.

tolbertam committed:
Make HostConnectionPoolTest more resilient

Use ConditionChecker to validate pool size as it is not guaranteed that
the connection has been established within 1 second so instead wait for
up to 5 seconds using the ConditionChecker.

Increase waiting for connection open attempt from 1 to 2s.

tolbertam commented:
Targeting for 3.0.3 as would be good to include getting this stable as part of the 3.0.3 test effort.

tolbertam commented:
tests using netty-tcnative for openssl broke on windows with 4.0.37 / Fork17.  Not sure why yet, will look into it.

tolbertam committed:
Update netty-tcnative to 1.1.33.Fork18

1.1.33.Fork17 does not work on windows.

tolbertam commented:
There are a few finicky tests that need to be solidified before i'd feel comfortable pulling this in and made part of PR checks.  I'll fix these and remove WIP once I feel confident.

tolbertam committed:
Set idempotence on Statement instead of globally

Makes test more reliable as its possible any ongoing queries could be
executed and increment the speculative execution counter.

Also reattempt test if scheduled execution did not occur.  It is
possible, yet unlikely, that a scheduled execution may not fire if the
timeout is not invoked before the response is received.   This should be
incredibly rare.

tolbertam commented:
Removing WIP tag, I think this is now stable enough to be deemed ready for review.   I will set it up to rebuild constantly to tease out any other tests we can solidify, but there are some that may be unavoidable (like CCM failing to start for whatever reason).

tolbertam committed:
Make test more resilient with higher timeout

tolbertam committed:
Check row is not null before other assertions

olim7t commented on driver-core/src/test/java/com/datastax/driver/core/ProtocolV1Test.java:
```diff
@@ -24,7 +25,7 @@
  * Tests targeting protocol v1 specifically.
  */
-@CCMConfig(version = "1.2.19")
+@CassandraVersion(major = 1.2, exactRequired = true)
```
So `exactRequired` skips the test if major.minor > 1.2?

tolbertam replied:
Effectively yes.   It skips any cassandra versions that doesn't match the criteria exactly.  i.e. if `major` is 1.2, only C\* versions matching 1.2.X will qualify.   If `major` is `1.2`, minor is `19`, only 1.2.19 will qualify.   I agree it's a bit crude.  @adutra has a good solution for this that allows a version range specification that I think will be much better/flexible that we can use in the future.

adutra replied:
So if I get it right, this test used to always be run against a 1.2.19 cluster, regardless of the default CCM version (`cassandra.version`), whereas now it would only run if the default CCM version _is_ 1.2.X, and would be skipped for all other configurations.

I'm fine with that, but indeed my feeling is that we should avoid skipping tests. IMHO there should be a default CCM version for the test suite that would be used for all tests that do not declare any specific CCM requirement, but if a test – like this one – has very specific requirements, such as a specific version or a specific version range, then a non-default CCM cluster should be spun for it, rather than simply skipping it. I basically came to this conclusion seeing how many times we realize very late that a test fails for a specific C\* version, just because the test was being skipped on our local machines.

In the new toolbox repository, things are being done that way (there is no test skipping). I understand that in a Jenkins matrix build that would make some tests be executed several times always against the same C\* version, but I'd rather have that, than risking to have some tests maybe being skipped everywhere.

tolbertam replied:
Right,  I just need a way of skipping this particular test since 1.2.x does not work well in CCM on windows.  I sort of agree with what you are saying though, should I find an alternative way for skipping this particular test on windows?

adutra replied:
Oh. So we need OS requirements too, I will look into that for the new test framework. 

Ideally, we would have to amend `@CCMConfig` to add a new attribute, e.g. `os = {NIX, MAC_OS}` and have the test skipped/ignored if the os doesn't match.

But that's complex, so in the meanwhile I think that what you currently have done so far is enough.

tolbertam replied:
I think to keep things really simple, I can just put a platform check in this test itself since I only need it for the 1 test.   This way we don't need to pollute `@CassandraVersion/@DseVersion/@CCMConfig` for the time being.  I'm sure we'll want it for more at some point, but by that time we'll be using the new test framework.

tolbertam replied:
Hrmm, unfortunately it doesn't seem like this is an option because there isn't a way for me prevent it creating the CCM cluster other than at annotation level, so may need to keep this in here for now.

adutra replied:
You need to hook in early enough to prevent the CCM cluster from being created. Try this:

``` java
    @Override
    public void beforeTestClass(Object testInstance) throws Exception {
        if (CCMBridge.isWindows())
            throw new SkipException("Only works on proper OS's");
        super.beforeTestClass(testInstance);
    }
```

(assuming `CCMBridge.isWindows()` is made accessible)

tolbertam replied:
Ah I see,  I added a `@BeforeClass` annotation and it got invoked after `CCMBridge` kicked in, but I see how overriding this could do the trick.  I'll do that.

tolbertam replied:
That worked great, thanks!  Pushed as 2d6c76c and deleted this commit.

olim7t commented on driver-core/src/test/java/com/datastax/driver/core/SchemaChangesTest.java:
```diff
@@ -482,8 +484,7 @@ public void should_notify_of_keyspace_drop(String keyspace) throws InterruptedEx
         refreshSchema();
-        for (Metadata m : metadatas())
-            assertThat(m.getReplicas(keyspace, Bytes.fromHexString("0xCAFEBABE"))).isNotEmpty();
+        replicasPresentForKeyspace(keyspace).becomesTrue();
```
Sorry but I'm not a big fan of this, kind of hard to read.

tolbertam replied:
Based on your comment below, I think I will need to separate out the validation between the true/false cases anyways.  Were you more against the naming strategy?  Or the gist of how the validation is done? (or both :P?)

tolbertam replied:
Come to think of it, @adutra has addressed all of this in #666, so i'll wait for that to merge and rebase and take in his changes.

tolbertam replied:
I deleted this change as it is no longer needed after #666 was merged.

olim7t commented on driver-core/src/test/java/com/datastax/driver/core/SchemaChangesTest.java:
```diff
@@ -500,8 +501,8 @@ public void should_notify_of_keyspace_drop(String keyspace) throws InterruptedEx
             assertThat(m.getKeyspace(keyspace)).isNull();
-            assertThat(m.getReplicas(keyspace, Bytes.fromHexString("0xCAFEBABE"))).isEmpty();
         }
+        replicasPresentForKeyspace(keyspace).becomesFalse();
```
The condition checks that the keyspace is present in the token map of all Clusters. So the negative would be "the keyspace is absent for at least one Cluster". Is it really what we want, or rather "the keyspace is absent for _all_ Clusters"?

tolbertam replied:
You are right, that doesn't seem like a complete validation.  I'll fix it.

tolbertam replied:
No longer applicable after #666.

tolbertam commented:
Something like ~7/85 of the builds I ran with this over the last week failed some of which I've addressed, so I think we may rarely see some intermittent failures, but I think this is tolerable for merge (after we got the other 3.0.3 tickets merged and this has been reviewed).

tolbertam commented:
Rebased against 3.0.x and deleted commit that made changes to `SchemaChangesTest` since it was no longer needed after #666.  Should be ready pending review and successful run.

adutra commented on driver-core/src/test/java/com/datastax/driver/core/utils/CassandraVersion.java:
```diff
@@ -34,7 +34,12 @@
+    /**
+     * @return Whether or not the exact major version is exactly required to run this test.
+     */
+    boolean[] exactRequired() default {false};
```
Why do you need this attribute to be a boolean array? It seems a simple boolean is enough.

tolbertam replied:
It's needed for defaulting the value if unspecified I believe.   The same thing is done in `@CCMConfig`

adutra replied:
`@CCMConfig` is complex in that it needs to distinguish `true`, `false` and `unset`, and so it uses boolean arrays = `{true}`, `{false}` and `{}`. But here I think you should only need:

``` java
boolean exactRequired() default false;
```

tolbertam replied:
Fixed and rebased commit 5bc921e (figured was simple enough to fix inline).

tolbertam committed:
Skip ProtocolV1Test on Windows

CCM does not work with C* 1.2 on windows and windows isn't officially
supported until 2.2.

adutra commented on driver-core/src/test/java/com/datastax/driver/core/CCMBridge.java:
```diff
@@ -876,8 +884,13 @@ public int weight() {
             for (String jvmArg : jvmArgs) {
-                allJvmArgs.append(" --jvm_arg=");
+                // Windows requires jvm arguments to be quoted, while *nix requires unquoted.
+                String quote = isWindows() ? "\"" : "";
```
This could be moved outside the for loop.

tolbertam replied:
thanks, will fix.

tolbertam replied:
Done, rebased it into 317b4d3 since it was a small change.
